python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -153,15 +176,86 @@ def caller(data, index, size):
     call_cuda_functions[funcname.lower()][type_key] = caller
 
 
-def _make_unary_attr(funcname):
-    class GroupUnaryReductionAttrTyping(AbstractTemplate):
-        key = f"GroupType.{funcname}"
+class GroupOpBase(AbstractTemplate):
+    def make_error_string(self, args):
+        fname = self.key.__name__
+        sr_err = ", ".join(["Series" for _ in range(len(args))])
+        return (
+            f"{fname}({sr_err}) is not supported by JIT GroupBy "
+            f"apply. Supported features are listed at:\n{_UDF_DOC_URL}"
+        )
+
+    def generic(self, args, kws):
+        # early exit to make sure typing doesn't fail for normal
+        # non-group ops
+        if not all(isinstance(arg, GroupType) for arg in args):
+            return None
+        # check if any groups are poisoned for this op
+        for arg in args:
+            if isinstance(arg.group_scalar_type, types.Poison):
+                raise UDFError(
+                    f"Use of a column of {arg.group_scalar_type.ty} detected "
+                    "within UDF body. \nOnly columns of the following dtypes "
+                    "may be used through the \nGroupBy.apply() JIT engine: "
+                    f"{[str(x) for x in SUPPORTED_GROUPBY_NUMPY_TYPES]}"
+                )
+        fname = self.key.__name__
+        if fname in call_cuda_functions:
+            for sig in call_cuda_functions[fname].keys():
+                if all(
+                    arg.group_scalar_type == ty for arg, ty in zip(args, sig)
+                ):
+                    return nb_signature(sig[0], *args)
+        raise UDFError(self.make_error_string(args))
+
+
+class GroupAttrBase(AbstractTemplate):
+    def make_error_string(self, args):
+        fname = self.key.split(".")[-1]
+        args = (self.this, *args)
+        dtype_err = ", ".join([str(g.group_scalar_type) for g in args])
+        sr_err = ", ".join(["Series" for _ in range(len(args) - 1)])
+        return (
+            f"\nSeries.{fname}({sr_err}) is not supported for "
+            f"({dtype_err}) within JIT GroupBy apply.\nTo see "
+            f"what's available, visit {_UDF_DOC_URL}"
+        )
 
-        def generic(self, args, kws):
-            for retty, inputty in call_cuda_functions[funcname.lower()].keys():
-                if self.this.group_scalar_type == inputty:
-                    return nb_signature(retty, recvr=self.this)
+    def generic(self, args, kws):
```

Comment:
Probably not necessary right now since there's only two, but if you start adding more of these types that are just error case we should try to abstract the common code into either a common parent class or some standalone helper functions to call so that the error checking and messages are as standardized as possible.

Reply:
Agree. I spent a little time trying to do this for a while, specifically having one base class for handling errors both for non-attr ops and attrs as well. The issue I ran into with it was that I couldn't make it simpler than computing a boolean `_IS_ATTR` up front and then branching `if`s based off it everywhere. In the end it felt like a class that was meant to do two separate things and it led me back to keeping it separate. But it's certainly possible that I just didn't boil the pot long enough :)